### PR TITLE
some initial code for app deployment controller/webhook

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
added code dealing with fetching component from application

### CODE REVIEW

File: pkg/controller/core.oam.dev/v1alpha2/applicationdeployment/applicationdeployment_controller.go [hunk 1]
```diff
@@ -32,26 +35,68 @@ type Reconciler struct {
...
+		klog.InfoS("get the source workload we need to work on", "sourceWorkload", klog.KObj(sourceWorkload))
+	}
+	// TODO: pass these two object to the rollout plan
 	return ctrl.Result{}, nil
 }
 
+func (r *Reconciler) handleFinalizer(appDeploy *corev1alpha2.ApplicationDeployment) {
+	if appDeploy.DeletionTimestamp.IsZero() {
+		if !slice.ContainsString(appDeploy.Finalizers, appDeployfinalizer, nil) {
+			// TODO: add finalizer
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/applicationdeployment/applicationdeployment_controller.go:85]: do you mean remove finalizer

File: pkg/controller/core.oam.dev/v1alpha2/applicationdeployment/applicationdeployment_controller.go [hunk 2]
```diff
@@ -32,26 +35,68 @@ type Reconciler struct {
...
+	}
+	klog.InfoS("get the target workload we need to work on", "targetWorkload", klog.KObj(targetWorkload))
+	if sourceWorkload != nil {
+		klog.InfoS("get the source workload we need to work on", "sourceWorkload", klog.KObj(sourceWorkload))
+	}
+	// TODO: pass these two object to the rollout plan
 	return ctrl.Result{}, nil
 }
 
+func (r *Reconciler) handleFinalizer(appDeploy *corev1alpha2.ApplicationDeployment) {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/applicationdeployment/applicationdeployment_controller.go:82]: what do you prepare to use finalizer for?

**Review by wonderflow — APPROVED** [review]

File: pkg/controller/core.oam.dev/v1alpha2/applicationdeployment/applicationdeployment_controller.go (hunk 1 above)
Comment (ryanzhang-oss) [reply pkg/controller/core.oam.dev/v1alpha2/applicationdeployment/applicationdeployment_controller.go:85]: this is the case that the object is not deleted (the delete timestamp is zero) and the finalizer is not there yet, so we need to add the finalizer there.

File: pkg/controller/core.oam.dev/v1alpha2/applicationdeployment/applicationdeployment_controller.go (hunk 2 above)
Comment (ryanzhang-oss) [reply pkg/controller/core.oam.dev/v1alpha2/applicationdeployment/applicationdeployment_controller.go:82]: I plan to make sure that any changes we make to the application is reverted

**Comment by wonderflow** [comment]
please resolve the conflict @ryanzhang-oss and we can merge this PR

**Pull request merged**